### PR TITLE
Expand API documentation of contract customization APIs.

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Contributes to #63686.

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/DefaultJsonTypeInfoResolver.cs [hunk 1]
```diff
@@ -35,7 +38,13 @@ private DefaultJsonTypeInfoResolver(bool mutable)
             s_defaultSimpleConverters ??= GetDefaultSimpleConverters();
         }
 
-        /// <inheritdoc/>
+        /// <summary>
+        /// Resolves a reflection-derived JSON contract for a given <paramref name="type"/> and <paramref name="options"/> configuration.
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/DefaultJsonTypeInfoResolver.cs:42]: This is virtual.  Is this guaranteed?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs [hunk 2]
```diff
@@ -31,8 +31,22 @@ public abstract partial class JsonTypeInfo
         private Action<object>? _onDeserialized;
 
         /// <summary>
-        /// Object constructor. If set to null type is not deserializable.
+        /// Gets or sets a parameterless constructor to be used on deserialization.
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs:34]: I find it confusing that this is called a "constructor".  I suggest using "factory" or something like that.  These aren't ctors in the .NET sense.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs [hunk 3]
```diff
@@ -165,6 +225,11 @@ public IList<JsonPropertyInfo> Properties
             {
                 VerifyMutable();
 
+                if (Kind == JsonTypeInfoKind.None)
+                {
+                    ThrowHelper.ThrowInvalidOperationException_JsonTypeInfoOperationNotPossibleForKind(Kind);
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs:230]: You added an exception to the xml docs, but does that map to this?  It doesn't seem so.

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/DefaultJsonTypeInfoResolver.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/DefaultJsonTypeInfoResolver.cs:42]: "this" == "reflection-derived"

I suggest the summary should just say `Resolves a JSON contract for a given <paramref name="type"/> and <paramref name="options"/> configuration.`  A `<remarks>...</remarks>` section can further provide commentary on the behavior of the base implementation, noting that a derived implementation may do something completely different.

**Comment by eiriktsarpalis** [comment]
CI failures caused by https://github.com/dotnet/runtime/issues/72429

**Pull request merged**